aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -38,13 +48,178 @@ __device__ __forceinline__ void load_store(
+      out = reinterpret_cast<T*>(small_buffer + offsets[idx]);
+    } else {
+      out = reinterpret_cast<T*>(buffer_ptr + offsets[idx]);
+    }
+  }
+};
+
+template <typename T>
+void pack_vectors_helper(
+    std::vector<const void*>& ptrs,
+    std::vector<size_t>& sizes,
+    std::vector<size_t>& offsets,
+    size_t& total_bytes,
+    const std::vector<T>& vec) {
+  ptrs.push_back(vec.data());
+  // Align the offset by sizeof(T)
+  total_bytes = at::round_up(total_bytes, sizeof(T));
+  offsets.push_back(total_bytes);
+  sizes.push_back(sizeof(T) * vec.size());
+  total_bytes += sizeof(T) * vec.size();
+}
+
+// Pack multiple vectors into a DevArrayPack
+//
+// When the total size of the input vectors is large, the function packs the
+// vectors into a device buffer allocated with CUDACachingAllocator. Thanks to
+// the allocator's stream awareness, the buffer can be safely accessed by the
+// kernel as long as the tensor owning the buffer is alive at the time of the
+// kernel launch. This tensor is returned to the caller, who is responsible for
+// keeping it alive until the kernel is launched.
+template <int n, typename... Vectors>
+std::tuple<DevArrayPack, c10::optional<at::Tensor>> pack_vectors(
+    const at::Device& device,
+    std::vector<const void*> (&addresses)[n],
+    const Vectors&... vecs) {
+  std::vector<const void*> ptrs;
+  std::vector<size_t> sizes;
+  std::vector<size_t> offsets;
+  size_t total_bytes = 0;
+
+  for (auto d = 0; d < n; ++d) {
+    pack_vectors_helper(ptrs, sizes, offsets, total_bytes, addresses[d]);
+  }
+  (pack_vectors_helper(ptrs, sizes, offsets, total_bytes, vecs), ...);
+
+  TORCH_CHECK(ptrs.size() <= DevArrayPack::max_arrays);
+  DevArrayPack pack{};
+
+  // Use the small buffer in DevArrayPack to pack the vectors
+  if (total_bytes < DevArrayPack::small_buffer_size) {
+    pack.buffer_ptr = nullptr;
+    for (const auto i : c10::irange(ptrs.size())) {
+      pack.offsets[i] = offsets[i];
+      memcpy(pack.small_buffer + offsets[i], ptrs[i], sizes[i]);
+    }
+    return std::make_tuple(pack, c10::optional<at::Tensor>(c10::nullopt));
+  }
+
+  bool is_capturing = at::cuda::currentStreamCaptureStatusMayInitCtx() !=
+      at::cuda::CaptureStatus::None;
+
+  // Use dynamically allocated buffer to pack the vectors
+  // First, stage the buffer content in host memory
+  auto buf_tensor = at::empty(
+      {static_cast<int64_t>(total_bytes)},
+      // Only use pinned memory when not capturing
+      at::TensorOptions().dtype(at::kByte).pinned_memory(!is_capturing));
+  for (const auto i : c10::irange(ptrs.size())) {
+    pack.offsets[i] = offsets[i];
+    memcpy(buf_tensor.data_ptr<uint8_t>() + offsets[i], ptrs[i], sizes[i]);
+  }
+
+  if (is_capturing) {
+    // When capturing, use a constant buffer to hold the packed vectors
+    at::cuda::get_current_capturing_graph()->alloc_const_buffer(
+        reinterpret_cast<void**>(&pack.buffer_ptr), total_bytes);
+    C10_CUDA_CHECK(cudaMemcpy(
```

Comment:
Sorry not totally following - why would this not work with an allocation from the cuda caching allocator ?

Reply:
Hmm.. now that you mention it, I may have prematurely assumed that it wouldn't work.

By using `CUDACachingAllocator::raw_alloc`, I reserve a buffer from the graph's private pool. By allowing it to "leak", the buffer won't be reassigned for any other purpose, and its lifetime is tied to the `CUDAGraph` object. This seems to be exactly the behavior I want, i.e., a buffer allocated at capture time that persists throughout the lifetime of the `CUDAGraph` object.

Is my understanding correct?

Reply:
> the buffer won't be reassigned for any other purpose, and its lifetime is tied to the CUDAGraph object. This seems to be exactly the behavior I want, i.e., a buffer allocated at capture time that persists throughout the lifetime of the CUDAGraph object

Right, this already the semantics of when you allocate from the cuda caching allocator during cuda graph capture